processing/src/main/java/io/druid/query/groupby/epinephelinae/BufferArrayGrouper.java

```diff
@@ -0,0 +1,281 @@
+  {
+    Preconditions.checkArgument(
+        dimIndex >= 0 && dimIndex < cardinalityWithMissingValue,
+        "Invalid dimIndex[%s]",
+        dimIndex
+    );
+
+    final ByteBuffer fromKey = checkAndGetKeyBuffer(keySerde, key);
+    if (fromKey == null) {
+      // This may just trigger a spill and get ignored, which is ok. If it bubbles up to the user, the message will
+      // be correct.
+      return Groupers.DICTIONARY_FULL;
+    }
+
+    final int recordOffset = dimIndex * recordSize;
+
+    if (recordOffset + recordSize > valBuffer.capacity()) {
+      // This error cannot be recoverd, and the query must fail
+      throw new ISE(
+          "A record of size [%d] cannot be written to the array buffer at offset[%d] "
+          + "because it exceeds the buffer capacity[%d]. Try increasing druid.processing.buffer.sizeBytes",
+          recordSize,
+          recordOffset,
+          valBuffer.capacity()
+      );
+    }
+
+    if (!isUsedSlot(dimIndex)) {
+      initializeSlot(dimIndex);
+    }
+
+    for (int i = 0; i < aggregators.length; i++) {
+      aggregators[i].aggregate(valBuffer, recordOffset + aggregatorOffsets[i]);
+    }
+
+    return AggregateResult.ok();
+  }
+
+  private ByteBuffer checkAndGetKeyBuffer(KeySerde<KeyType> keySerde, KeyType key)
+  {
+    final ByteBuffer fromKey = keySerde.toByteBuffer(key);
+    if (fromKey == null) {
+      return null;
+    }
+
+    if (fromKey.remaining() != keySerde.keySize()) {
+      throw new IAE(
+          "keySerde.toByteBuffer(key).remaining[%s] != keySerde.keySize[%s], buffer was the wrong size?!",
+          fromKey.remaining(),
+          keySerde.keySize()
+      );
+    }
+    return fromKey;
+  }
+
+  private void initializeSlot(int dimIndex)
+  {
+    final int index = dimIndex / 8;
+    final int extraIndex = dimIndex % 8;
+    usedFlagBuffer.put(index, (byte) (usedFlagBuffer.get(index) | 1 << extraIndex));
+
+    final int recordOffset = dimIndex * recordSize;
+    for (int i = 0; i < aggregators.length; i++) {
+      aggregators[i].init(valBuffer, recordOffset + aggregatorOffsets[i]);
+    }
+  }
+
+  private boolean isUsedSlot(int dimIndex)
+  {
+    final int index = dimIndex / 8;
+    final int extraIndex = dimIndex % 8;
+    final int usedByte = 1 << extraIndex;
+    return (usedFlagBuffer.get(index) & usedByte) == usedByte;
+  }
+
+  @Override
+  public void reset()
+  {
+    final int usedBufferCapacity = usedFlagBuffer.capacity();
+    for (int i = 0; i < usedBufferCapacity; i++) {
+      usedFlagBuffer.put(i, (byte) 0);
+    }
+  }
+
+  @Override
+  public void close()
+  {
+    for (BufferAggregator aggregator : aggregators) {
+      try {
+        aggregator.close();
+      }
+      catch (Exception e) {
+        LOG.warn(e, "Could not close aggregator [%s], skipping.", aggregator);
+      }
+    }
+  }
+
+  @Override
+  public Iterator<Entry<KeyType>> iterator(boolean sorted)
+  {
+    // result is always natually sorted by keys
```

Comment:
My bad. I think it would be better to throw UnsupportedOperationException if sorted is true for now. This can be removed when we use BufferArrayGrouper for merging partial aggregates.